source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc

```diff
@@ -18,7 +18,72 @@ namespace Extensions {
 namespace HttpFilters {
 namespace HeaderToMetadataFilter {
 
-Rule::Rule(const std::string& header, const ProtoRule& rule) : header_(header), rule_(rule) {
+// extract the value of the header
+absl::optional<std::string> HeaderValueSelector::extract(Http::HeaderMap& map) const {
+  const Http::HeaderEntry* header_entry = map.get(header_);
+  if (header_entry == nullptr) {
+    return absl::optional<std::string>(); // no value
+  }
+  // catch the value in the header
+  absl::optional<std::string> value = std::string(header_entry->value().getStringView());
+  if (remove_) {
+    map.remove(header_); // remove the header
+  }
+  return value;
+}
+
+// extract the value of the key from the cookie header
+absl::optional<std::string> CookieValueSelector::extract(Http::HeaderMap& map) const {
+  std::string value = Envoy::Http::Utility::parseCookieValue(map, cookie_);
+  if (!value.empty()) {
+    return absl::optional<std::string>(value);
+  }
+  return absl::optional<std::string>(); // no value
+}
+
+Rule::Rule(const ProtoRule& rule) : rule_(rule) {
+  const auto& option = rule.header_cookie_specifier_case();
+
+  // initialize the shared pointer
+  switch(option) {
```

Comment:
As mentioned above, we'll need to just directly check `rule.header()` and `rule.cookie()` for non-empty values, with a TODO similar to the one below re: v4.